src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,300 @@
+
+from aspire.utils import ensure
+from aspire.utils.random import Random
+
+
+class Rotation:
+    def __init__(
+        self, num_rots, seed=None, angles=None, matrices=None, dtype=np.float32
+    ):
+        """
+         Initialize a Rotation object
+
+         :param num_rots: Total number of rotation sets
+         :param seed: Seed for initializing random rotations.
+             If None, the random generator is not re-seeded.
+         :param angles: Euler angles (in radian) to generate rotation matrices.
+             If None, Predefined rotation matrices or uniformly
+             distributed angles will be used.
+        :param matrices: Rotation matrices to initialize Rotation object.
+             If None, Predefined or uniformly distributed angles
+             will be used.
+         :param dtype: data type for angles and rotation matrices
+        """
+        self.dtype = np.dtype(dtype)
+        self.num_rots = None
+        self.angles = None
+        self.matrices = None
+        self.shape = None
+        if angles is not None:
+            ensure(
+                num_rots == angles.shape[0],
+                "Number of rotation matrices should equal to "
+                "number of sets of Euler angles.",
+            )
+            self.from_euler(angles, dtype=dtype)
+        elif matrices is not None:
+            ensure(
+                num_rots == matrices.shape[0],
+                "Number of rotation matrices should equal to "
+                "number of input sets of matrices.",
+            )
+            self.from_matrix(matrices, dtype=dtype)
+        else:
+            self.num_rots = num_rots
+            angles = Rotation.uniform_random_angles(num_rots, seed=seed, dtype=dtype)
+            self.from_euler(angles, dtype=dtype)
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return f"Rotation stack consisting of {self.num_rots} elements of {self.dtype} type"
+
+    def __len__(self):
+        return self.num_rots
+
+    def __getitem__(self, item):
+        return self.matrices[item]
+
+    def __setitem__(self, key, value):
+        self.matrices[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            output = self.matrices @ other.matrices
+        else:
+            output = self.matrices @ other
+        return output
+
+    @property
+    def T(self):
+        """
+        Shortcut to get full set of transposed rotation matrices
+        """
+        return self.invert()
+
+    def invert(self):
+        """
+        Apply transpose operation to all rotation matrices
+
+        :return: The set of transposed matrices
+        """
+        return np.transpose(self.matrices, axes=(0, 2, 1))
+
+    def asnumpy(self):
+        """
+        :return: Rotation matrices as a n x 3 x 3 array
+        """
+        return self.matrices
+
+    def from_euler(self, values, dtype=np.float32):
```

Comment:
The two `from_` functions should be static. Also need one for randomly distributed rotations.

Reply:
Changed to use static property in new commit a6c2f53.